### PR TITLE
fix(4746): Do not link and sync pkgs that fail to download

### PR DESCRIPTION
**Pull request by m-czernek**
## What does this PR change?

When a package fails to download, `reposync.py` still tries to process and link it internally. This PR fixes https://github.com/uyuni-project/uyuni/issues/4746

## GUI diff

No difference.

Before:

After:

- [ ] **DONE**

## Documentation
- No documentation needed: only internal and user invisible changes

- [ ] **DONE**

## Test coverage
- No tests: **add explanation**
- No tests: already covered
- Unit tests were added
- Cucumber tests were added

- [x] **DONE**

## Links

Fixes #
Tracks # **add downstream PR, if any**

- [ ] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: python/spacewalk/satellite_tools/download.py [hunk 1]
```diff
@@ -282,6 +284,12 @@ def run(self):
                 # log_obj must be thread-safe
                 self.parent.log_obj.log(success, os.path.basename(params['relative_path']))
             self.queue.task_done()
+            if not success:
+                self.lock.acquire()
```
Comment (m-czernek) [inline python/spacewalk/satellite_tools/download.py:288]: Not totally sure if these locks are necessary in the threads themselves

File: python/spacewalk/satellite_tools/reposync.py [hunk 2]
```diff
@@ -1152,6 +1152,9 @@ def import_packages(self, plug, source_id, url, is_non_local_repo):
         downloader.set_log_obj(logger)
         downloader.run()
 
+        log(0, 'Filtering packages that failed to download')
+        to_process = [i for i in to_process if i[0].getNVREA() not in downloader.failed_pkgs]
```
Comment (agraul) [inline python/spacewalk/satellite_tools/reposync.py:1156]: Worst case this is `O(n^2)`, right? `downloader.failed_pkgs` will in most cases be a short list so this does not matter much, but a set could also make sense for `downloader.failed_pkgs`

**Review by agraul — COMMENTED** [review]
I'm still trying to follow the multithreading logic, we might not need to lock these.

File: python/spacewalk/satellite_tools/reposync.py (hunk 2 above)
Comment (m-czernek) [reply python/spacewalk/satellite_tools/reposync.py:1156]: Makes sense; it was my assumption that the list of failed pkgs won't be long, but you're totally right about the set.

File: python/spacewalk/satellite_tools/download.py [hunk 3]
```diff
@@ -374,6 +383,12 @@ def run(self):
                 while any(t.is_alive() for t in started_threads):
                     time.sleep(1)
                 break
+            # accumulate all failed packages, if any
+            failed_pkgs = [pkg for t in started_threads for pkg in t.failed_pkgs]
+            if bool(failed_pkgs):
+                self.lock.acquire()
+                self.failed_pkgs = failed_pkgs
+                self.lock.release()
```
[the remark this answers is not in the document]
Comment (m-czernek) [reply python/spacewalk/satellite_tools/download.py]: I have the same intuition, though I wasn't totally sure I'm not missing something. Removed the thread locking

File: python/spacewalk/satellite_tools/download.py [hunk 4]
```diff
@@ -282,6 +283,10 @@ def run(self):
                 # log_obj must be thread-safe
                 self.parent.log_obj.log(success, os.path.basename(params['relative_path']))
             self.queue.task_done()
+            if not success:
+                package = os.path.basename(params['relative_path'])
+                # append package name without extension
+                self.failed_pkgs.add(package.rsplit('.', maxsplit=1)[0])
```
Comment (vzhestkov) [inline python/spacewalk/satellite_tools/download.py:289]: It makes me a bit confused. Actually afterwards it's checking with NVREA (strange by itself as usually it's called NEVRA as Epoch has the higher priority over the Version, but anywat) the potential problem here, that some of the third party repos are not folliwing the rule to call the package files with this naming convention. For exmple: https://rpm.opera.com/rpm. It contains `opera_stable-103.0.4928.26-linux-release-x64-signed.rpm` with `NEVRA`=`google-chrome-stable-118.0.5993.70-1.x86_64`

Comment (m-czernek) [reply python/spacewalk/satellite_tools/download.py]: Nice, makes sense! Now, instead of relying on the package naming rules, I rely on paths. This means that there's no need for removing the extension anymore, and from my understanding, the target path should always be defined.

WDYT about the current version?

File: python/test/unit/spacewalk/satellite_tools/test_reposync.py [hunk 5]
```diff
@@ -212,6 +212,63 @@ def test_sync_success_regen(self):
...
+        for i, pack in enumerate(packs):
+            pack.arch = "arch1"
+            pack.checksum = ""
+            # The first package failed to download
+            pack.unique_id.relativepath = fail_pkg_name if i == 0 else "pkg"
+
+        plug =  self._mock_repo_plugin(packs)
+
+        with patch("uyuni.common.context_managers.CFG", self._mock_cfg()), \
+            patch.object(spacewalk.satellite_tools.reposync.RepoSync, "chunks", Mock(return_value=())):
```
Comment (agraul) [inline python/test/unit/spacewalk/satellite_tools/test_reposync.py:264]: What exactly is this doing? Why is it needed?

File: python/test/unit/spacewalk/satellite_tools/test_reposync.py [hunk 6]
```diff
@@ -212,6 +212,63 @@ def test_sync_success_regen(self):
...
+            rs.import_packages(plug, None, "unused-url-string", None)
+
+        to_process = apply_async_mock.call_args_list[0][1]["args"][0]
+        self.assertTrue(len(to_process) == 1)
+
+        pack_to_process = to_process[0][0]
+        self.assertTrue(pack_to_process.unique_id.relativepath != fail_pkg_name)
```
Comment (agraul) [inline python/test/unit/spacewalk/satellite_tools/test_reposync.py:271]: Do I read this correctly that first `to_process` is set to a "test data" defined in this test case and then the assertions are done on `to_process`?

File: python/test/unit/spacewalk/satellite_tools/test_reposync.py [hunk 7]
```diff
@@ -212,6 +212,63 @@ def test_sync_success_regen(self):
...
+    @patch("uyuni.common.context_managers.initCFG", Mock())
+    @patch("spacewalk.satellite_tools.reposync.log2", Mock())
+    @patch("spacewalk.satellite_tools.reposync.os", os)
+    @patch("spacewalk.satellite_tools.reposync.ThreadedDownloader")
+    @patch("spacewalk.satellite_tools.reposync.multiprocessing.Pool")
+    def test_import_packages_excludes_failed_pkgs(self, pool, downloader):
+        """
+        When downloader fails to download a subset of packages
+        Then the RepoSync.import_packages function should not process the failed packages
+        """
```
Comment (agraul) [inline python/test/unit/spacewalk/satellite_tools/test_reposync.py:240]: I find this test a bit hard to follow. It's not obvious what's the system under test (SUT) and what are the collaborators (that can be substituted by test doubles). Could you maybe re-structure this test a bit and maybe add a few comments?

**Review by agraul — CHANGES_REQUESTED** [review]
Thanks for adding tests! I know it's not easy with our reposync test suite, but you've done a good job!

I have left a few comments about a few details, once those are resolved we can clean up the git history and merge!

File: python/test/unit/spacewalk/satellite_tools/test_reposync.py (hunk 7 above)
Comment (m-czernek) [reply python/test/unit/spacewalk/satellite_tools/test_reposync.py:253]: It is absolutely hard to follow. Honestly, the function under test is hard to follow as well :/ It does multiple things with multiple objects that we do not care about in this particular test, hence the number of patches.

I have added a few comments and refactored the test to make it just a little bit less noisy. However, the main collaborators should be objects injected by pytest into the method parameter list, right?

File: python/test/unit/spacewalk/satellite_tools/test_download.py [hunk 8]
```diff
@@ -117,3 +145,34 @@ def test_reposync_download_thread_fetch_url_proxy_pass():
         assert "proxy" not in url_grabber_opts_mock.mock_calls[1].kwargs
         assert "username" not in url_grabber_opts_mock.mock_calls[1].kwargs
         assert "password" not in url_grabber_opts_mock.mock_calls[1].kwargs
+
+
+def test_reposync_download_thread_sets_failed_pkg():
```
[the remark this answers is not in the document]
Comment (agraul) [reply python/test/unit/spacewalk/satellite_tools/test_download.py:150]: For me it's okay to keep 1) The tests execute quickly enough that the execution time is not impacted noticeably while 1) can help pinpointing the source of problems. :+1:

File: python/test/unit/spacewalk/satellite_tools/test_reposync.py (hunk 7 above)
Comment (agraul) [reply python/test/unit/spacewalk/satellite_tools/test_reposync.py:253]: You're totally right, `import_packages` is doing too much.

Thanks for adding the comments, I think the test is better now. It's not pretty but following the function, it's understandable.

File: python/test/unit/spacewalk/satellite_tools/test_reposync.py (hunk 5 above)
Comment (agraul) [reply python/test/unit/spacewalk/satellite_tools/test_reposync.py]: The reason that `SQLBackend` and `ChannelPackageSubscription` need to be mocked is that there is a package that "was downloaded", right? If so, I get it now.

File: python/test/unit/spacewalk/satellite_tools/test_reposync.py (hunk 6 above)
Comment (agraul) [reply python/test/unit/spacewalk/satellite_tools/test_reposync.py]: The names are fine. I was mostly confused why `to_process` is taken from `apply_async_mock.call_args_list[0][1]["args"][0]`.

File: python/test/unit/spacewalk/satellite_tools/test_reposync.py (hunk 5 above)
Comment (m-czernek) [reply python/test/unit/spacewalk/satellite_tools/test_reposync.py]: As per our discussions, simplified test -> removed this mock.

File: python/test/unit/spacewalk/satellite_tools/test_reposync.py (hunk 6 above)
Comment (m-czernek) [reply python/test/unit/spacewalk/satellite_tools/test_reposync.py]: As per our discussion, simplified the test -> removed this call.

File: python/spacewalk/satellite_tools/download.py (hunk 4 above)
Comment (vzhestkov) [reply python/spacewalk/satellite_tools/download.py]: Thanks for fixing it :+1: Looks good for me.

**Review by vzhestkov — APPROVED** [review]
Looks good :+1: 
Thanks for the fix.

**Review by agraul — APPROVED** [review]
LGTM! :rocket:
